server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java

```diff
@@ -624,11 +654,31 @@ public MemoryUsage overLimit(MemoryUsage memoryUsed) {
                         blackHole += localBlackHole;
                         logger.trace("black hole [{}]", blackHole);
 
-                        long now = timeSupplier.getAsLong();
-                        this.lastCheckTime = now;
-                        allocationDuration = now - begin;
+                        this.lastCheckTime = timeSupplier.getAsLong();
                         this.attemptNo = 0;
                     }
+
+                    if (gcTriggeredSuccessfully == false) {
+                        long now = timeSupplier.getAsLong();
+                        var canPerformFullGC = now >= lastFullGCTime + fullGCMinimumInterval;
+                        if (canPerformFullGC) {
+                            // Enough time passed between 2 full GC fallbacks
+                            // Increase the lock timeout: threads coming here to this method after we decided a full GC is needed will be
+                            // able to wait for longer. Threads already holding the lock will wait the normal timeout.
+                            currentLockTimeout = fullGCLockTimeout;
```

Comment:
I don't get how this is supposed to work! ;-) Since already waiting threads will not see it, as your comment suggests. Should threads not able to acquire the lock check the currentLockTimeout value and retry the lock if appropriate ?

Reply:
There are several options here.
In the current state, already waiting threads will just wait the "normal" timeout. New threads reaching the lock after this point (but before the potentially longer part of the operation, the full GC) will wait the extended timeout. It's a compromise.
Like you pointed out, a possible solution is that threads that fail the wait can retry it, if a full GC is undergoing, with an extended timeout. 
Or we change the wait to always be a loop and divide it into chunks (e.g. make the lock timeout 500 -> 50 and try it 10 times in a loop).
Or do not perform the GC in the current call, update the timeout and book it for the next iteration (but actually this could also have threads waiting with the "normal" timeout)

Reply:
Changed to retry to acquire the lock (with a longer timeout)